src/section/section_map.rs

```diff
@@ -0,0 +1,461 @@
+
+            let old_key_info = self
+                .keys
+                .remove(&old_pfx)
+                .expect("Bug in BTreeMap for update_keys");
+
+            self.recent_keys.push_front((old_pfx, old_key_info.clone()));
+            if self.recent_keys.len() > MAX_RECENT_KEYS {
+                let _ = self.recent_keys.pop_back();
+            }
+
+            trace!("    from {:?} to {:?}", old_key_info, key_info);
+
+            let old_pfx_sibling = old_pfx.sibling();
+            let mut current_pfx = key_info.prefix().sibling();
+            while !self.keys.contains_key(&current_pfx) && current_pfx != old_pfx_sibling {
+                let _ = self.keys.insert(current_pfx, old_key_info.clone());
+                current_pfx = current_pfx.popped().sibling();
+            }
+        }
+        let _ = self.keys.insert(*key_info.prefix(), key_info.clone());
+    }
+
+    pub fn get_knowledge(&self, prefix: &Prefix<XorName>) -> Option<u64> {
+        self.knowledge.get(prefix).copied()
+    }
+
+    /// Returns the index of the public key in our_history that will be trusted by the target
+    /// location
+    pub fn proving_index(&self, target: &DstLocation) -> u64 {
+        let (prefix, &index) = if let Some(pair) = self
+            .knowledge
+            .iter()
+            .filter(|(prefix, _)| target.is_compatible(prefix))
+            .min_by_key(|(_, index)| *index)
+        {
+            pair
+        } else {
+            return 0;
+        };
+
+        if let Some(sibling_index) = self.knowledge.get(&prefix.sibling()) {
+            // The sibling section might not have processed the split yet, so it might still be in
+            // `target`'s location. Because of that, we need to return index that would be trusted
+            // by them too.
+            index.min(*sibling_index)
+        } else {
+            index
+        }
+    }
+
+    /// Provide a start index of a SectionProofSlice that proves the given signature to the given
+    /// destination location.
+    /// If `node_knowledge_override` is `Some`, it is used when calculating proof for
+    /// `DstLocation::Node` instead of the stored knowledge. Has no effect for other location types.
+    pub fn knowledge_index(
+        &self,
+        target: &DstLocation,
+        node_knowledge_override: Option<u64>,
+    ) -> u64 {
+        match (target, node_knowledge_override) {
+            (DstLocation::Node(_), Some(knowledge)) => knowledge,
+            _ => self.proving_index(target),
+        }
+    }
+
+    /// Updates the entry in `knowledge` for `prefix` to the `version`; if a split
+    /// occurred in the meantime, the versions for sections covering the rest of the address space
+    /// are initialised to the old version that was stored for their common ancestor
+    pub fn update_knowledge(&mut self, prefix: Prefix<XorName>, version: u64) {
+        trace!(
+            "attempts to update knowledge of our elders_info with version {:?} for \
```

Comment:
I think this comment is bit misleading.
The `knowledge` is our perspective that other section think what is our version is.
So, I think the comment shall better be:
("Section (prefix {:?}) shall now know about our section version {:?}", prefix, version)